hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/HoodieMergeHelper.java

```diff
@@ -72,93 +70,116 @@ public static HoodieMergeHelper newInstance() {
-      if (writeInternalSchema.isEmptySchema()) {
-        throw new HoodieException(String.format("cannot find file schema for current commit %s", commitInstantTime));
-      }
-      List<String> colNamesFromQuerySchema = querySchema.getAllColsFullName();
-      List<String> colNamesFromWriteSchema = writeInternalSchema.getAllColsFullName();
-      List<String> sameCols = colNamesFromWriteSchema.stream()
-              .filter(f -> colNamesFromQuerySchema.contains(f)
-                      && writeInternalSchema.findIdByName(f) == querySchema.findIdByName(f)
-                      && writeInternalSchema.findIdByName(f) != -1
-                      && writeInternalSchema.findType(writeInternalSchema.findIdByName(f)).equals(querySchema.findType(writeInternalSchema.findIdByName(f)))).collect(Collectors.toList());
-      readSchema = AvroInternalSchemaConverter
-          .convert(new InternalSchemaMerger(writeInternalSchema, querySchema, true, false, false).mergeSchema(), readSchema.getName());
-      Schema writeSchemaFromFile = AvroInternalSchemaConverter.convert(writeInternalSchema, readSchema.getName());
-      needToReWriteRecord = sameCols.size() != colNamesFromWriteSchema.size()
-              || SchemaCompatibility.checkReaderWriterCompatibility(readSchema, writeSchemaFromFile).getType() == org.apache.avro.SchemaCompatibility.SchemaCompatibilityType.COMPATIBLE;
-      if (needToReWriteRecord) {
-        renameCols = InternalSchemaUtils.collectRenameCols(writeInternalSchema, querySchema);
-      }
-    }
+    // In case Advanced Schema Evolution is enabled we might need to rewrite currently
+    // persisted records to adhere to an evolved schema
+    Option<Function<GenericRecord, GenericRecord>> schemaEvolutionTransformerOpt =
+        composeSchemaEvolutionTransformer(writerSchema, baseFile, writeConfig, table.getMetaClient());
+
+    // Check whether the writer schema is simply a projection of the file's one, ie
+    //   - Its field-set is a proper subset (of the reader schema)
+    //   - There's no schema evolution transformation necessary
+    boolean isPureProjection = isProjectionOf(readerSchema, writerSchema)
+        && !schemaEvolutionTransformerOpt.isPresent();
+    // Check whether we will need to rewrite target (already merged) records into the
+    // writer's schema
+    boolean shouldRewriteInWriterSchema = writeConfig.shouldUseExternalSchemaTransformation()
+        || !isPureProjection
+        || baseFile.getBootstrapBaseFile().isPresent();
+
+    HoodieExecutor<GenericRecord, GenericRecord, Void> wrapper = null;
 
     try {
-      final Iterator<GenericRecord> readerIterator;
+      Iterator<GenericRecord> recordIterator;
+
+      // In case writer's schema is simply a projection of the reader's one we can read
+      // the records in the projected schema directly
+      ClosableIterator<GenericRecord> baseFileRecordIterator =
+          reader.getRecordIterator(isPureProjection ? writerSchema : readerSchema);
```

Comment:
can we add docs on whats reader schema referring to and whats writer schema referring to.
wrt reading one avro file eg, usual nomenclature is, writer schema refers to the schema w/ which the file was written, and reader schema refers to the new schema w/ which we are trying to read the file back. 

here we have 3 things in play. 
we are reading an existing file using schema Y which could have been written using schema X, and we are passing it to UpsertPartitioner w/ possible schema evolution. 

So, is schema X the writer schema and schema Y refers to reader schema? 
may be we should add docs whereever we deal w/ both these schema and maintain uniformity across the code.

Reply:
I didn't add the docs, just b/c now these vars are immutable and are easily traceable (L81-82).

In this context:
 - Reader is the schema with which we're reading the existing file
 - Writer schema is the schema with which we're gonna be writing files

Reply:
if we maintain the same terminology, we are good. 
thought we might have some confusion in HoodieMergeHandler, since we read from one file and write into another file.  guess the read from old file is handled w/n MergeHelper.

Reply:
Oh, i see.
There was previously a confusion, and it was the primary reason we had issues with it (table vs write schema), now it's in line with the standard terms (writer-schema for writing, reader-schema for reading)